dbms/src/Storages/Transaction/SSTReader.h

```diff
@@ -14,28 +14,110 @@
 
 #pragma once
 
+#include <Common/Logger.h>
 #include <Common/nocopyable.h>
 #include <Storages/Transaction/ProxyFFI.h>
 
 namespace DB
 {
-
 struct SSTReader
 {
-    bool remained() const;
-    BaseBuffView key() const;
-    BaseBuffView value() const;
-    void next();
+    virtual bool remained() const;
+    virtual BaseBuffView key() const;
+    virtual BaseBuffView value() const;
+    virtual void next();
 
     DISALLOW_COPY_AND_MOVE(SSTReader);
     SSTReader(const TiFlashRaftProxyHelper * proxy_helper_, SSTView view);
-    ~SSTReader();
+    // Should only be used when SSTReader is a base class.
+    SSTReader();
+    virtual ~SSTReader();
 
 private:
     const TiFlashRaftProxyHelper * proxy_helper;
     SSTReaderPtr inner;
     ColumnFamilyType type;
+    bool inited;
 };
 
+/// MultiSSTReader helps when there are multiple sst files in a column family.
+/// It is derived from virtual class SSTReader, so it can be holded in a SSTReaderPtr.
+/// It also maintains instance of `R` which is normaly SSTReader(and MockSSTReader in tests),
+/// to read a single sst file.
+/// An `Initer` function is need to create a instance of `R` from a instance of `E`,
+/// which is usually path of the SST file.
+/// We introduce `R` and `E` to make this class testable.
+template <typename R, typename E>
+struct MultiSSTReader : SSTReader
```

Comment:
```suggestion
struct MultiSSTReader : public SSTReader
```

Reply:
Why public here?

Reply:
It is `Private Inheritance` if you ignore "public" here. Why do you need to make it a private inheritance?

Reply:
Inheritance is public by default for structs, the following compiles
```C++
struct B {
    int z = 1;
};

struct D : B {

};

int main() {
    B* p = new D();
    printf("%d\n", p->z);
}
```

Reply:
Well... I didn't know that struct is public inheritance by default before. No matter what, I think explicit is better than implicit.
And as SSTReader now has virtual methods, all its members are private, I think it is more likely to be a `class` instead of `struct`, though they are almost the same in C++. So does `MultiSSTReader`

Reply:
Change from struct to class makes sense.